lib/shared/puppet_self_service.rb

```diff
@@ -124,4 +124,153 @@ def self.filesystem_free(path)
+    else
+      result.error
+    end
+  end
+
+  # Execute psql queries.
+  #
+  # @param sql [String] SQL to execute via a psql command.
+  # @param psql_options [String] list of options to pass to the psql command.
+  #
+  # @return (see #exec_return_result)
+
+  def self.psql_return_result(sql, psql_options = '')
+    command = %(su pe-postgres --shell /bin/bash --command "cd /tmp && #{PUP_PATHS[:server_bin]}/psql #{psql_options} --command \\"#{sql}\\"")
+    exec_return_result(command)
+  end
+
+  # Below method to execute the PSQL statement to identify thundering herd
+  def self.psql_thundering_herd
+    sql = %(
+      SELECT date_part('month', start_time) AS month,
+      date_part('day', start_time) AS day,
+      date_part('hour', start_time) AS hour,
+      date_part('minute', start_time) as minute, count(*)
+      FROM reports
+      WHERE start_time BETWEEN now() - interval '7 days' AND now()
+      GROUP BY date_part('month', start_time), date_part('day', start_time), date_part('hour', start_time), date_part('minute', start_time)
+      ORDER BY date_part('month', start_time) DESC, date_part('day', start_time) DESC, date_part( 'hour', start_time ) DESC, date_part('minute', start_time) DESC;
+    )
+    psql_options = '--dbname pe-puppetdb'
+    psql_return_result(sql, psql_options)
+  end
+end
+
+# Execute external commands
+module Exec
+  # Command Result
+  #
+  # @param stdout [String] A string containing the standard output
+  #   written by the command.
+  # @param stderr [String] A string containing the standard error
+  #   written by the command.
+  # @param status [Integer] An integer representing the exit code of the
+  #   command.
+  # @param error [String, nil] A string holding an error message if
+  #   command execution was not successful.
+  Result = Struct.new(:stdout, :stderr, :status, :error)
+
+  # Exception class for failed commands
+  class ExecError < StandardError; end
+
+  # Execute a command and return a Result
+  #
+  # This is basically `Open3.popen3`, but with added logic to time the
+  # executed command out if it runs for too long.
+  #
+  # @param cmd [Array<String>] Command and arguments to execute. Commands
+  #  consisting of a single String will cause `Process.spawn` to wrap the
+  #  execution in a system shell such as `/bin/sh -c`.
+  # @param env [Hash] A hash of environment variables to set
+  #   when the command is executed.
+  # @param stdin_data [String] A string of standard input to pass
+  #   to the executed command. Sould be smaller than 4096 characters.
+  # @param timeout [Integer] Number of seconds to allow for command
+  #   execution to complete. A value of 0 will disable the timeout.
+  #
+  # @return [Result] A `Result` object containing the output and status
+  #   of the command.
+  def self.exec_cmd(*cmd, env: {}, stdin_data: nil, timeout: 300)
+    out_r, out_w = IO.pipe
+    err_r, err_w = IO.pipe
+    _env = { 'LC_ALL' => 'C', 'LANG' => 'C' }.merge(env)
```

Comment:
Looks like ruby does not like variable beginning with an underscore so we need to change this